helix-core/src/main/java/org/apache/helix/messaging/handling/HelixTaskExecutor.java

```diff
@@ -361,17 +374,13 @@ ExecutorService findExecutorServiceForMsg(Message message) {
         executorService = _batchMessageExecutorService;
       } else {
         Message.MessageInfo msgInfo = new Message.MessageInfo(message);
-        String perResourceTypeKey =
-            msgInfo.getMessageIdentifier(Message.MessageInfo.MessageIdentifierBase.PER_RESOURCE);
-        String perStateTransitionTypeKey =
-            msgInfo.getMessageIdentifier(Message.MessageInfo.MessageIdentifierBase.PER_STATE_TRANSITION_TYPE);
-        if (perStateTransitionTypeKey != null && _executorMap.containsKey(perStateTransitionTypeKey)) {
-          LOG.info(String.format("Find per state transition type thread pool for resource %s from %s to %s",
-              message.getResourceName(), message.getFromState(), message.getToState()));
-          executorService = _executorMap.get(perStateTransitionTypeKey);
-        } else if (_executorMap.containsKey(perResourceTypeKey)) {
-          LOG.info("Find per-resource thread pool with key: " + perResourceTypeKey);
-          executorService = _executorMap.get(perResourceTypeKey);
+        for (int i = Message.MessageInfo.MessageIdentifierBase.values().length - 1; i >= 0; i--) {
+          String msgIdentifer = msgInfo.getMessageIdentifier(Message.MessageInfo.MessageIdentifierBase.values()[i]);
+          if (msgIdentifer != null && _executorMap.containsKey(msgIdentifer)) {
+            LOG.info(String.format("Find customized threadpool for %s", msgIdentifer));
+            executorService = _executorMap.get(msgIdentifer);
+            break;
```

Comment:
In the Enum we declare types generally from coarser to finer, so we search in reverse order. If a finer type has a threadpool assigned, then we use the finer one.

Reply:
Oh I see, I think declare from lower level to higher level is more intuitive, so I will keep it this way.

Reply:
+1 on this. And we could use:
`for (Message.MessageInfo.MessageIdentifierBase value : Message.MessageInfo.MessageIdentifierBase.values())`
for simpler code and this value can be used in line 380 as well.

Reply:
Discussed with @xyuanlu offline. Changing enum order does simplify code, but users will not see this block of code because this is our internal logic. What users see is the enum declaration when they use the API, so making the enum easy to understand for users is more important. Therefore, keeping this order.
The original comments I put besides each enum declaration are a bit confusing in terms of their hierarchy, updated them.